src/client/api/json_parser/payloads.c

```diff
@@ -353,15 +353,15 @@ cJSON* json_tagged_serialize(tagged_data_t* tagged_data) {
 
   // tag
   if (tagged_data->tag) {
-    if (!cJSON_AddStringToObject(tagged_data_payload, JSON_KEY_TAG, (const char* const)tagged_data->tag->data)) {
-      printf("[%s:%d] add tag type failed\n", __func__, __LINE__);
+    char tag_str[TAGGED_DATA_TAG_MAX_LENGTH_BYTES] = {0};
+    if (bin_2_hex(tagged_data->tag->data, tagged_data->tag->len, tag_str,
+                  BIN_TO_HEX_STR_BYTES(tagged_data->tag->len)) != 0) {
```

Comment:
Probably a `tag` won't always be a string. It could just contain some pure binary data. It's up to the user to put NULL terminator at the end of `tag` array if data inside represents a string.
So `BIN_TO_HEX_STR_BYTES(tagged_data->tag->len)` should be replaced with `BIN_TO_HEX_BYTES(tagged_data->tag->len)`.
Maximum `tag` binary data length is 64 bytes. Maximum `tag` hex string length is 128 bytes.

Reply:
1. tag_str[TAGGED_DATA_TAG_MAX_LENGTH_BYTES] - > tag_str[BIN_TO_HEX_STR_BYTES(TAGGED_DATA_TAG_MAX_LENGTH_BYTES)].

2. bin_2_hex needs a buffer of size 2 * input binary len + 1, as it appends '\0' at the end.